Mvp 76

hojkim77 opened the pull request:
# 🔍 What is this PR for? | PR 설명

> PR 을 만들게 된 원인 / 문제상황 / 해결법 등을 설명해주세요.
### 리뷰 관련
- 리뷰 관련 api 작성(Create, Read)
- 리뷰 정렬 기능 추가
### (보호자 입장)예약 관련
- 예약 객체의 전역적인 관리
- 예약 객체의 리뷰, 예약 상태등 변경에대한 리페칭 문제 해결

# 📝 Changes | 핵심 변경사항

> 리뷰어가 알아야 할 핵심 변경사항이 있다면 미리 말해주세요.

- 리뷰 관련 api가 추가되었습니다. 이는 스크린과 연결되었습니다.
- (보호자 입장)예약 관련 model(CLBooking, CLBookingStore)이 작성되었습니다. 이는 아직 사용중이지 않습니다.

# 📷 Screenshots | 스크린샷

> PR 을 설명할 스크린샷을 첨부할 수 있다면, 첨부해주세요.

-

# 🤦🏻‍♂️ Help me | 도와주세요

> 해결하지 못한 점이 있어, 도움이 필요한 경우 남겨주세요.

- 예약 관련 모델에 대한 피드백이 필요합니다.

hojkim77 committed:
MVP-76: (진행중) 리뷰 조회 api 작성 및 연결

hojkim77 committed:
MVP-76: 리뷰 작성 api 작성 및 연결 - 데이터 페칭 문제로인해 model 사용 필요

hojkim77 committed:
MVP-76: 리뷰 조회 api 작성 및 연결 완료

hojkim77 committed:
MVP-76: 리뷰 정렬 기능 추가

hojkim77 committed:
MVP-73: (진행중)보호자 입장 예약 모델 작성

hojkim77 committed:
feat: 예약 및 예약Store 모델 작성

smnchoi commented on app/models/cl-booking-store/cl-booking-store.test.ts:
@hojkim77 
`*.test.ts` 파일은 우선 다 삭제해주세요 :)

smnchoi commented on app/models/cl-booking-store/cl-booking-store.ts:
```diff
@@ -0,0 +1,64 @@
+  .actions(withSetPropAction)
+  .views((self) => ({
+    getBooking(bookingId) {
+      const targetBooking = self.watingBookings.find(
```
@hojkim77 
1. `targetBooking`의 값이 없을 경우 (undefined) 에는 어떻게 되나요? 드문 케이스 이지만, 기존에 가지고 있던 MST 데이터가 손실되었고, 그 상태에서 getBooking 이 실행된다면, 앱이 충돌합니다. 예외처리 구문을 추가해주세요.
2. `self.watingBookings` 가 undefined 일 경우에는 .find 메서드를 사용할 수 없습니다. `self.watingBookings?.find` 로 수정해주세요.

smnchoi commented on app/models/cl-booking-store/cl-booking-store.ts:
```diff
@@ -0,0 +1,64 @@
+    getBooking(bookingId) {
+      const targetBooking = self.watingBookings.find(
+        (booking) => booking.visitingBookingId === bookingId,
+      ) as ClBooking
```
`as` 키워드는 신중히 사용해야 합니다.
참고: https://www.phind.com/search?cache=yfysul25rqlto946y7f8awag

강제로 type assertion 을 구현하기 떄문에, 위에서 제가 피드백드린 사항을 고민해 볼 만한 여지조차 잃을 수 있습니다.
지금 상황에서는 `as` 키워드를 쓰지 않고도 타입스크리트 에러 메시지를 없앨 수 있을 것 같습니다. 수정해주세요.

smnchoi commented on app/models/cl-booking/cl-booking.ts:
```diff
@@ -0,0 +1,39 @@
+    profileImage: types.string,
+    desc: types.string,
+    status: types.frozen<BookingStatus>(),
+    ratings: types.optional(types.number, 0),
```
@hojkim77 더 간단하게 수정해주세요:
```ts
ratings: 0
...
isFavorite: false
```

smnchoi commented on app/screens/_CLIENT/booking-stack/past-bookings/past-bookings-screen.tsx:
```diff
@@ -16,15 +17,21 @@ export const PastBookingsScreen: FC<
     navigation.setParams(undefined)
   }, [])
 
   const fetchBookings = useCallback(async () => {
```
@hojkim77 `fetchBookings()` 함수는 `useFocusEffect` 내부에서만 쓰이네요. 지금 상황에서는 이곳에서 정의해서 사용하는 것은 복잡도만 증가시킵니다. 이번 PR 에서는 이 함수의 로직을`useFocusEffect` 내부로 옮기는게 더 좋아 보입니다.

smnchoi commented on app/screens/_CLIENT/caregiver-detail-stack/all-reviews-screen/all-reviews-screen.tsx:
```diff
@@ -1,25 +1,68 @@
 
-  useLayoutEffect(() => {
-    setReviews(_reviews)
+  useEffect(() => {
```
@hojkim77 어떠한 이유때문에 `useLayoutEffect` 으로 바꾸셨나요?

hojkim77 replied:
useLayoutEffect의 남용이라 생각되어 useEffect로 수정했습니다. 다시 확인 부탁드립니다!

smnchoi replied:
제가 잘못 읽었군요 :)

smnchoi commented on app/services/api/review.ts:
```diff
@@ -112,12 +113,15 @@ export const uploadURIS = async (images: PickerImage[]): Promise<string[] | []>
-export const postVisitingReview = async (params: PostReviewParams): Promise<boolean> => {
+export const postVisitingReview = async (
+  userId: number,
+  params: PostReviewParams,
```
@hojkim77 변수명을 다음과 같이 수정해주세요:
`params ` -> `body`
`PostReviewParams` -> `PostVisitingReviewRequestBody`

REST API Request 에서, params 와 body 에 차이는 다음 글을 참고하세요:
https://velog.io/@montoseon/param-vs-query-vs-body

smnchoi commented on app/services/api/review.ts:
```diff
@@ -112,12 +113,15 @@ export const uploadURIS = async (images: PickerImage[]): Promise<string[] | []>
     // * 서버에 이미지를 upload 하는 과정
     const postParams: PostReviewToServerParams = {
-      userId: USER_ID,
+      userId: userId,
```
@hojkim77 
그냥 프로퍼티 네임에 쓰이는 변수명이랑, 프로퍼티 밸류에 쓰이는 변수명이 동일한 경우,
`userId: userId`, -> `userId,` 으로만 적으셔도 됩니다. (ES6 문법)

smnchoi commented on app/services/api/review.ts:
```diff
@@ -227,3 +239,66 @@ export const getCrecheReview = async (bookingId: number): Promise<Review | null>
+      return { isSuccess: false }
+    }
+
+    return { isSuccess: true, visitingReviews: response.data.visitingReviews }
```
@hojkim77 
`response.data.visitingReviews` 가 `null` 혹은 `undefined` 일 가능성은 없나요?

getVisitingReviews 호출 이후에,
```ts
 const [reviews, setReviews] = useState<visitingReview[]>([])
```
사용되는 reviews 의 기본값이 빈 배열 [] 이므로,

프로그램이 의도대로 작동하기 위해서는, 만약에 null 일 가능성이 존재한다면,  빈 배열 [] 을 리턴하도록 해야 합니다.
스웨거 API 문서 확인 및, 실제 API 가 동작하는 리턴값을 확인하고,
만약 이 API 의 response 가 null 혹은 undefined 를 반환할 가능성이 존재한다면,
다음과 같이 수정해주세요:
```
return { isSuccess: true, visitingReviews: response.data.visitingReviews ?? [] }
```
---
[참고) `??` nullish coalescing operator](https://stackoverflow.com/questions/61480993/when-should-i-use-nullish-coalescing-vs-logical-or)

minSsan commented on app/models/cl-booking-store/cl-booking-store.ts:
```diff
@@ -0,0 +1,64 @@
+     */
+    waitingsToCurrents(bookingId: number) {
+      const targetBooking = self.getBooking(bookingId)
+      self.watingBookings.remove(targetBooking)
```
[수민님이 남겨주신 리뷰](https://github.com/Care-Giver/CareGiver/pull/204/files#r1535215188)대로 `getBooking` 내에서 **optinal chaining**으로 예외처리를 한다면, `getBooking`의 리턴 값 또한 `undefined`가 될 수 있으므로 이에 대한 예외 처리도 필요할 것 같습니다 ..!

minSsan commented on app/models/cl-booking-store/cl-booking-store.ts:
```diff
@@ -0,0 +1,64 @@
+     * 특정 예약 리뷰 상태를 "Complete"로 변경합니다.
+     */
+    setCompleteReview(bookingId: number) {
+      const targetBooking = self.getBooking(bookingId)
```
위의 리뷰와 동일합니다

hojkim77 committed:
fix: PR리뷰 수정사항 반영

smnchoi commented on app/models/cl-booking-store/cl-booking-store.ts:
```diff
@@ -16,12 +16,40 @@ export const ClBookingStoreModel = types
+        return targetBooking
+      } catch (error) {
+        console.error("[cl-booking-store getBooking Error!] >>>", error)
+        return false
```
@hojkim77 
getBooking 메서드는 성공시, ClBooking 객체를 반환합니다.
실패시, false 대신, null 을 반환하도록 수정해주세요.

성공시 true 를 반환하는 메서드일 경우에만, (실패시) false 를 반환하도록 해주세요.

smnchoi commented on app/services/api/review.ts:
```diff
@@ -110,30 +110,30 @@ export const uploadURIS = async (images: PickerImage[]): Promise<string[] | []>
     }
 
-    const AwsUris = await uploadURIS(params.images)
+    const AwsUris = await uploadURIS(body.images)
```
@hojkim77 
"클래스"가 아니라면, 모든 변수는 소문자로 시작해야 합니다. (`camelCase`)

smnchoi commented on app/models/cl-booking-store/cl-booking-store.ts:
```diff
@@ -36,24 +64,28 @@ export const ClBookingStoreModel = types
-      self.watingBookings.remove(targetBooking)
-      self.currentBookings.push(targetBooking)
+      const targetBooking = self.getBooking(bookingId, BookingStatus.WAITING)
+      if (targetBooking) {
```
@hojkim77 
"Guard Clause" 는 정상적인 흐름에 대해 if 를 사용하는 것이 아니라, _예외 상황을 if - return 으로 처리하는 것입니다._
```ts
// 예외
if(!targetBooking) return
// 정상 로직
self.watingBookings.remove(targetBooking)
self.currentBookings.push(targetBooking)
```
위와 같은 방식으로 전부 수정해주세요.

https://velog.io/@swhan9404/%ED%81%B4%EB%A6%B0%EC%BD%94%EB%93%9C%EB%A5%BC-%EC%9C%84%ED%95%9C-Guard-Clause

hojkim77 committed:
fix: PR리뷰 수정사항 반영_2

smnchoi committed:
[후기 작성 기능 추가, UT 관련 더미데이터 삭제] Merge branch 'release/1.0.0' into MVP-76

# Conflicts:
#	app/components/review-box/review-box.tsx
#	app/models/index.ts
#	app/screens/_CLIENT/booking-stack/all-bookings-screen/all-bookings-screen.tsx
#	app/screens/_CLIENT/caregiver-detail-stack/all-reviews-screen/all-reviews-screen.tsx
#	app/screens/_CLIENT/caregiver-detail-stack/caregiver-detail-information-screen/caregiver-detail-information-screen.tsx

smnchoi committed:
fix: setSelectedComment state 로직 개선

smnchoi committed:
fix(댓글 수정): commentId 제대로 설정하도록 수정 - 기존에는 commentId 에 작성자 id 가 잘못 할당되어 있었음

smnchoi committed:
fix: __commentator__ 에 optional chain 추가, updateComment 에러메시지 수정

smnchoi committed:
fix(댓글 작성/수정): debounce 추가, disabled 추가

smnchoi committed:
fix(댓글 스크린 헤더): 작성 or 수정 헤더 텍스트 수정

smnchoi committed:
fix(주로 사용중인 버튼 컴포넌트들): debounce 추가
